% This one will format for two-sided binding (ie left and right pages have mirror margins; blank pages inserted where needed):
% \documentclass[a4paper,12pt]{ociamthesis}
% This one will format for one-sided binding (ie left margin > right margin; no extra blank pages):
%\documentclass[a4paper]{ociamthesis}
% This one will format for PDF output (ie equal margins, no extra blank pages):
\documentclass[a4paper,nobind]{ociamthesis} 
\pdfminorversion=7



%%%%% SELECT YOUR DRAFT OPTIONS
% Three options going on here; use in any combination.  But remember to turn the first two off before
% generating a PDF to send to the printer!

% This adds a "DRAFT" footer to every normal page.  (The first page of each chapter is not a "normal" page.)
% \fancyfoot[C]{\emph{DRAFT Printed on \today}}  

% This highlights (in blue) corrections marked with (for words) \mccorrect{blah} or (for whole
% paragraphs) \begin{mccorrection} . . . \end{mccorrection}.  This can be useful for sending a PDF of
% your corrected thesis to your examiners for review.  Turn it off, and the blue disappears.
\correctionstrue

\usepackage{subfigure}
\usepackage{listings}
\usepackage{amssymb}
\usepackage{longtable}
\usepackage{array}


%%%%% BIBLIOGRAPHY SETUP
% The science-type option: numerical in-text citation with references in order of appearance.
\usepackage[style=numeric-comp, sorting=none, backend=biber, doi=false, isbn=false, language=british]{biblatex}
\newcommand*{\bibtitle}{References}

% This makes the bibliography left-aligned (not 'justified') and slightly smaller font.
\renewcommand*{\bibfont}{\raggedright\small}

% Change this to the name of your .bib file (usually exported from a citation manager like Zotero or EndNote).
\addbibresource{references.bib}
% graphics path
\graphicspath{ {./figures/} }


% Uncomment this if you want equation numbers per section (2.3.12), instead of per chapter (2.18):
%\numberwithin{equation}{subsection}



%%%%% THESIS / TITLE PAGE INFORMATION
% Everybody needs to complete the following:
\title{Predicting Prime Editing Outcome with Machine Learning}
\author{Peiheng Lu}
\college{Wolfson College}

% Master's candidates who require the alternate title page (with candidate number and word count)
% must also un-comment and complete the following three lines:
\masterssubmissiontrue
\candidateno{1592800}
\wordcount{17,263}

% Uncomment the following line if your degree also includes exams (eg most masters):
%\renewcommand{\submittedtext}{Submitted in partial completion of the}
% Your full degree name.  (But remember that DPhils aren't "in" anything.  They're just DPhils.)
\degree{MSc Advanced Computer Science}
% Term and year of submission, or date if your board requires (eg most masters)
\degreedate{Michaelmas 2024}


%%%%% YOUR OWN PERSONAL MACROS
% This is a good place to dump your own LaTeX macros as they come up.

% To make text superscripts shortcuts
\renewcommand{\th}{\textsuperscript{th}} % ex: I won 4\th place
\newcommand{\nd}{\textsuperscript{nd}}
\renewcommand{\st}{\textsuperscript{st}}
\newcommand{\rd}{\textsuperscript{rd}}

%%%%% THE ACTUAL DOCUMENT STARTS HERE
\begin{document}

%%%%% CHOOSE YOUR LINE SPACING HERE
% This is the official option.  Use it for your submission copy and library copy:
\setlength{\textbaselineskip}{22pt plus2pt}
% This is closer spacing (about 1.5-spaced) that you might prefer for your personal copies:
%\setlength{\textbaselineskip}{18pt plus2pt minus1pt}

% You can set the spacing here for the roman-numbered pages (acknowledgements, table of contents, etc.)
\setlength{\frontmatterbaselineskip}{18pt plus2pt minus1pt}

% Leave this line alone; it gets things started for the real document.
\setlength{\baselineskip}{\textbaselineskip}

%%%%% CHOOSE YOUR SECTION NUMBERING DEPTH HERE
% You have two choices.  First, how far down are sections numbered?  (Below that, they're named but
% don't get numbers.)  Second, what level of section appears in the table of contents?  These don't have
% to match: you can have numbered sections that don't show up in the ToC, or unnumbered sections that
% do.  Throughout, 0 = chapter; 1 = section; 2 = subsection; 3 = subsubsection, 4 = paragraph...

% The level that gets a number:
\setcounter{secnumdepth}{2}
% The level that shows up in the ToC:
\setcounter{tocdepth}{2}


%%%%% ABSTRACT SEPARATE
% This is used to create the separate, one-page abstract that you are required to hand into the Exam
% Schools.  You can comment it out to generate a PDF for printing or whatnot.
% \begin{abstractseparate}

% \end{abstractseparate}


% JEM: Pages are roman numbered from here, though page numbers are invisible until ToC.  This is in
% keeping with most typesetting conventions.
\begin{romanpages}

% JEM: By default, this template uses the traditional Oxford "Belt Crest". Un-comment the following
% line to use the newer, "Blue Square" logo:
\renewcommand{\crest}{{\includegraphics[width=4.2cm, height=4.2cm]{figures/newlogo.pdf}}}

% Title page is created here
\maketitle

%%%%% DEDICATION -- If you'd like one, un-comment the following.
%\begin{dedication}
%This thesis is dedicated to\\
%someone\\
%for some special reason\\
%\end{dedication}

%%%%% ACKNOWLEDGEMENTS -- Nothing to do here except comment out if you don't want it.
\begin{acknowledgements}

    Thank you to my supervisor, and soon to be Dr. (censored for anonymity, I'll make sure you see the full version after mark is released). Your guidance and support have been invaluable throughout this project, especially when I am navigating the unfamiliar territories of deep learning and gene editing. Taking up this project with no deep learning experience was a daunting task, but your patience and encouragement have made the journey much more manageable. Regardless of the outcome, I have learned a lot from this project, and I am grateful for the opportunity to work with you.

    I would also like to thank my friends for help me going through the rough patches of the project, the experience would have been much more grueling without your support. The everyday cooking and chatting have been a great way to unwind after a long day of work, and I am grateful for the friendship we have built over the past months. 

    Finally, at the (possible) end of my student life, I would like to thank my family for their unwavering support and encouragement. Supporting me through my education abroad has been a significant financial and emotional burden, but they never mentioned it once through the five years of my study. They have always been my biggest cheerleaders, and would like nothing in return but my happiness and success.
\end{acknowledgements}

%%%%% ABSTRACT -- Nothing to do here except comment out if you don't want it.
\begin{abstract}

    Prime editing is one of the latest advances in the field of gene editing, offering a much more versatile and precise way to edit the genome compared to the CRISPR-Cas9 and base editor systems. However, its application is significantly limited by low efficiencies when using unoptimized guide RNAs. To solve this problem, a number of computational tools have been developed for in silico guide optimization, and achieved varying degrees of success. 

    This study further enriched the toolkits of computational guide optimization by developing a number of new machine learning models that offer superior or comparable performance to existing tools on a number of datasets. 

    The project began with the development of a benchmarking dataset of prime editing outcomes from the literatures, including a total of 23 datasets from 2 different studies. Five of these datasets have a large number of examples (over 20,000) and were used as the main datasets for model training and evaluation, while the rest served as additional datasets for fine tuning the deep learning models and improve their practical usability.

    A number of ensemble models were developed, leveraging classical ensemble methods including weighted averaging, bagging and AdaBoost. The ensembles contained the existing PRIDICT and DeepPrime models, as well as other more traditional machine learning algorithms, including XGBoost and ridge regression. 

    Additionally, the transformer architecture, shown to be highly effective in sequence-to-sequence tasks, was also explored for prime editing efficiency prediction. However, it failed to achieve the same level of performance as the ensemble models on the main datasets. This could be due to the inherent differences between the tasks of prime editing and base editing efficiency prediction, where the architecture previously found success.

    The best performing model was presented as a web application, allowing users without deep learning expertise to input their target edits and receive a number of recommended pegRNA designs.
\end{abstract}

%%%%% MINI TABLES
% This lays the groundwork for per-chapter, mini tables of contents.  Comment the following line
% (and remove \minitoc from the chapter files) if you don't want this.  Un-comment either of the
% next two lines if you want a per-chapter list of figures or tables.
\dominitoc % include a mini table of contents
%\dominilof  % include a mini list of figures
%\dominilot  % include a mini list of tables

% This aligns the bottom of the text of each page.  It generally makes things look better.
\flushbottom

% This is where the whole-document ToC appears:
% This is where the whole-document ToC appears:
\tableofcontents

\listoffigures
\mtcaddchapter
% \mtcaddchapter is needed when adding a non-chapter (but chapter-like) entity to avoid confusing minitoc

% Uncomment to generate a list of tables:
%\listoftables
%	\mtcaddchapter

%%%%% LIST OF ABBREVIATIONS
% This example includes a list of abbreviations.
\include{text/abbreviations}
\end{romanpages}

\input{text/introduction.tex}
\input{text/method.tex}
\input{text/benchmarking.tex}
\input{text/discussion.tex}

\startappendices
% Add or remove any appendices you'd like here:


%%%%% REFERENCES


{\renewcommand*\MakeUppercase[1]{#1}%

% \input{text/appendix-models.tex}
\input{text/appendix.tex}


\printbibliography[heading=bibintoc,title={\bibtitle}]}


\end{document}